fn/option.go

```diff
@@ -0,0 +1,149 @@
+package fn
+
+// Option[A] represents a value which may or may not be there. This is very
+// often preferable to nil-able pointers.
+type Option[A any] struct {
+	isSome bool
+	some   A
+}
+
+// Some trivially injects a value into an optional context.
+//
+// Some : A -> Option[A].
+func Some[A any](a A) Option[A] {
+	return Option[A]{
+		isSome: true,
+		some:   a,
+	}
+}
+
+// None trivially constructs an empty option
+//
+// None : Option[A].
+func None[A any]() Option[A] {
+	return Option[A]{}
+}
+
+// ElimOption is the universal Option eliminator. It can be used to safely
+// handle all possible values inside the Option by supplying two continuations.
+//
+// ElimOption : (Option[A], () -> B, A -> B) -> B.
+func ElimOption[A, B any](o Option[A], b func() B, f func(A) B) B {
```

Comment:
Not familiar with the naming convention but to me this rather sounds like it does unwrap the value and calls the function modifier which transforms the type to B. So perhaps could be `MapUnwrap` (or `UnwrapMap`)? Eliminating feels like it wouldn't return anything but rather swallow the value (of type A).

Reply:
I don't have any attachments to naming personally. I named it based off of the [formalism](https://en.wikipedia.org/wiki/Disjunction_elimination#:~:text=has%20to%20be%20true.,Q%2C%20Q%20is%20certainly%20true) it simulates.

Reply:
Looks like the Option stuff got merged as part of #8026